File: src/typescript/2022/02/p1.ts
const enum Shape {
    ROCK = 1,
    PAPER = 2,
    SCISSOR = 3
}

type OpponentShape = 'A' | 'B' | 'C';
type PlayerShape = 'X' | 'Y' | 'Z';

type Match = `${OpponentShape}${PlayerShape}`;

const opponentShapeMap: Record<OpponentShape, Shape> = {
    A: Shape.ROCK,
    B: Shape.PAPER,
    C: Shape.SCISSOR
}

const playerShapeMap: Record<PlayerShape, Shape> = {
    X: Shape.ROCK,
    Y: Shape.PAPER,
    Z: Shape.SCISSOR
}

const winningShapeMap: Record<Shape, Shape> = {
    [Shape.ROCK]: Shape.SCISSOR,
    [Shape.PAPER]: Shape.ROCK,
    [Shape.SCISSOR]: Shape.PAPER
}

const enum Outcome {
    WIN = 6,
    DRAW = 3,
    LOSE = 0
}

export function p1(input: string): number {
    const contents = input.split("\n").filter(x => x.length).map(x => x[0] + x[2]) as Match[];
    let points = 0;
    for (let i = 0; i < contents.length; i++) {
        const [op, pl] = contents[i] as unknown as [OpponentShape, PlayerShape];
        if (opponentShapeMap[op] === playerShapeMap[pl]) points += Outcome.DRAW;
        if (winningShapeMap[playerShapeMap[pl]] === opponentShapeMap[op]) points += Outcome.WIN;
        points += playerShapeMap[pl]
    }
    return points;
}

File: src/go/2022/02/p1.go
package p1

import (
	"errors"
	"fmt"
	"os"
	"strings"
)

const (
	ROCK int = iota + 1
	PAPER
	SCISSOR
)

func p1(input string) (*int, error) {
	if input == "" {
		return nil, errors.New("Input is empty")
	}
	contents := strings.Split(strings.ReplaceAll(strings.TrimSpace(input), " ", ""), "\n")
	winningCombination := map[int]int{ROCK: SCISSOR, PAPER: ROCK, SCISSOR: PAPER}
	Outcome := map[string]int{"WIN": 6, "DRAW": 3, "LOSE": 0}
	PlayerHands := map[string]int{"X": ROCK, "Y": PAPER, "Z": SCISSOR}
	OpponentHands := map[string]int{"A": ROCK, "B": PAPER, "C": SCISSOR}

	points := 0
	for _, item := range contents {
        var op, pl string = string(item[0]), string(item[1])
		if OpponentHands[op] == PlayerHands[pl] {
			points += Outcome["DRAW"]
		}
		if winningCombination[PlayerHands[pl]] == OpponentHands[op] {
			points += Outcome["WIN"]
		}
		points += PlayerHands[pl]
	}
	return &points, nil
}

func main() {
	file, err := os.ReadFile("p1.input")

	if err != nil {
		fmt.Println("err: ", err)
	}

	raw := string(file)

	fmt.Println(p1(raw))

}
